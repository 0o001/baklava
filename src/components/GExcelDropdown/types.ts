import { GExcelDropdownStatus } from '@/components/GExcelDropdown/enums';

export interface GExcelDropdownListItem {
  status: GExcelDropdownStatus,
  fileName: string,
  downloadUrl: string,
}
