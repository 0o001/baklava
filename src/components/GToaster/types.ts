type GToasterVariant = 'error' | 'success' | 'warning' | 'info';
type Duration = {
  seconds: number,
  autoClose: boolean,
  classes?: string[],
  style?: any,
  onCloseId?: ReturnType<typeof setTimeout>,
}
type GToasterDataConfig = {
  title: string,
  group: string,
  duration: number,
  autoClose: boolean,
  variants: any
}
export type CallbackType = 'onCloseCallback' | 'onShowCallback'

export interface GToasterOptions {
  id?: number;
  variant?: GToasterVariant;
  title?: string;
  unique?: boolean;
  group?: string;
  duration?: number;
  autoClose?: boolean;
  closeOnClick?: boolean;
  onCloseCallback?: (data: any) => void;
  textColor?: string;
  onShowCallback?: (data: any) => void;
  props?: any
}

export interface GToasterItem extends GToasterOptions {
  message?: string;
  durationObject: Duration;
  variantObject?: any;
  closed?: boolean
}

export interface GToasterData {
  items: GToasterItem[];
  id: number;
  config: GToasterDataConfig
}

export interface Toast {
  hide: (id: number) => void;
  hideAll: () => void;
  success: (message: string, toasterOptions?: GToasterOptions) => void;
  error: (message: string, toasterOptions?: GToasterOptions) => void;
  info: (message: string, toasterOptions?: GToasterOptions) => void;
  warning: (message: string, toasterOptions?: GToasterOptions) => void;
}
