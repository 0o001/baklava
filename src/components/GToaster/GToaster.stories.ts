// @ts-nocheck
import Vue from 'vue';
import { storiesOf } from '@storybook/vue';
import { withKnobs, text, boolean, number, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import GToasterPlugin from '../../plugins/GToaster.plugin';
import GToaster from '../GToaster';
import GButton from '../GButton';
import { GToasterOptions } from '@/components/GToaster/types';

const stories = storiesOf('GToaster', module);

stories.addDecorator(withKnobs);

Vue.use(GToasterPlugin, { GToaster });

stories
  .add('Show Defaults', () => ({
    template: `
      <div style="display: flex; flex-direction: column; width: 200px">
        <GButton @click="showSuccess" variant="success">Show Success</GButton>
        <GButton @click="showError" variant="danger">Show Error</GButton>
        <GButton @click="showWarning" variant="warning">Show Warning</GButton>
        <GButton @click="showInfo" variant="info">Show Info</GButton>
        <GButton @click="hideAll" class="btn btn-dark">Hide All</GButton>
     </div>
    `,
    components: {
      GButton,
    },
    methods: {
      showSuccess () {
        Vue.$toast.success();
      },
      showError (): void {
        // @ts-ignore
        this.$toast.error();
      },
      showWarning () {
        Vue.$toast.warning('Warning', {
          onCloseCallback: (message: any) => {
            action('On Close')(message);
          },
          onShowCallback: (message: any) => {
            action('On Show')(message);
          },
        });
      },
      showInfo () {
        // @ts-ignore
        this.$toast.info('Info');
      },
      hideAll () {
        // @ts-ignore
        this.$toast.hideAll();
      },
    },
    mounted () {
      const toastTypes = ['error', 'success', 'warning', 'info'];
      toastTypes.forEach(variant => {
        // @ts-ignore
        this.$toast[variant]('Message');
      });
    },
  }))
  .add('Show with options', () => ({
    props: {
      type: {
        default: select('Toaster Type', ['success', 'error', 'warning', 'info'], 'success'),
      },
      message: {
        default: text('Message', 'Message'),
      },
      title: {
        default: text('Title', 'Title'),
      },
      autoClose: {
        default: boolean('Auto Close (Related with Duration)', true),
      },
      duration: {
        default: number('Duration', 3000),
      },
      closeOnClick: {
        default: boolean('Close On Click', false),
      },
      unique: {
        default: boolean('Unique (Related with Group Name)', false),
      },
      group: {
        default: text('Group Name', 'Unique Name'),
      },
      textColor: {
        default: text('Text Color', 'text-white'),
      },
    },
    components: {
      GButton,
    },
    template: `
        <GButton @click="showMessage" class="btn btn-primary">Show Toaster</GButton>
    `,
    methods: {
      showMessage (): void {
        const options: GToasterOptions = {
          ...this.duration ? { duration: this.duration } : {},
          ...{ closeOnClick: this.closeOnClick },
          ...this.group ? { group: this.group } : {},
          ...{ unique: this.unique },
          ...{ autoClose: this.autoClose },
          ...{ title: this.title },
          ...{ textColor: this.textColor },
          onCloseCallback: (message) => {
            action('On Close')(message);
          },
          onShowCallback: (message) => {
            action('On Show')(message);
          },
        };

        // @ts-ignore
        this.$toast[this.type](this.message, options);
      },
    },
  }));
