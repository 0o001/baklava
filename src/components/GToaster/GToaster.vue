<template>
  <div class="toaster-wrapper">
    <transition-group name="fade">
      <div
        v-for="item in reversedItems"
        :key="item.id"
        :class="[item.variantObject.className]"
        class="toaster"
        @click="!item.closeOnClick || onClose(item)"
      >
        <div :class="[`${item.variantObject.className}_header`, 'header']">
          <div
            class="title"
            v-if="item.title"
          >
            <GText :color="item.textColor">
              {{ item.title }}
            </GText>
          </div>
          <div
            class="close"
            @click.self="onClose(item)"
          >
            Ã—
          </div>
        </div>
        <component
          :is="item.message"
          v-if="typeof item.message === 'object'"
          v-bind="item.props"
          class="content"
        />
        <div
          v-else
          class="content"
        >
          <GText :color="item.textColor">
            {{ item.message }}
          </GText>
        </div>
        <div
          v-if="item.durationObject.autoClose"
          class="duration-wrapper"
        >
          <div
            class="duration"
            :class="item.durationObject.classes"
            :style="item.durationObject.style"
          />
        </div>
      </div>
    </transition-group>
  </div>
</template>

<script lang="ts">
import { CallbackType, GToasterData, GToasterItem, GToasterOptions } from './types.ts';

import GText from '../GText/GText.vue';

export default {
  name: 'GToaster',

  components: {
    GText,
  },

  data (): GToasterData {
    const items: GToasterItem[] = [];

    return {
      items,
      id: 0,
      config: {
        title: '',
        group: 'default',
        duration: 2000,
        autoClose: false,
        variants: {
          error: {
            className: 'error',
          },
          success: {
            className: 'success',
          },
          warning: {
            className: 'warning',
          },
          info: {
            className: 'info',
          },
        },
      },
    };
  },

  computed: {
    reversedItems (): GToasterItem[] {
      return this.items.slice().reverse();
    },
  },

  beforeDestroy () {
    this.items.forEach((item: GToasterItem) => {
      if (item.durationObject && item.durationObject.onCloseId) {
        clearTimeout(item.durationObject.onCloseId);
      }
    });
  },

  methods: {
    toast (message: string, config: GToasterOptions = {}) {
      const { variants }: { variants: any } = this.config;
      const variant = variants[config.variant || ''] || variants.success;
      const title = config.title || this.config.title || variant.title;
      const id = config.id || this.id++;
      const { unique } = config || this.config;
      const group = config.group || this.config.group;
      const duration = config.duration || this.config.duration;
      const textColor = config.textColor || 'text-white';
      const autoClose = config.autoClose || this.config.autoClose ||
        config.duration !== undefined || this.config.duration !== 2000 || false;
      const closeOnClick = config.closeOnClick !== false;
      const props = config.props || null;
      const onCloseCallback = config.onCloseCallback;
      const onShowCallback = config.onShowCallback;

      if (unique && this.items.some(item => group !== 'default' && item.group === group)) {
        return;
      }

      const item: GToasterItem = {
        id,
        message,
        title,
        variantObject: variant,
        unique,
        group,
        textColor,
        durationObject: {
          seconds: duration,
          autoClose,
        },
        closeOnClick,
        props,
        onCloseCallback,
        onShowCallback,
        closed: false,
      };

      this.items.push(item);

      this.executeCallback(item, 'onShowCallback');

      if (autoClose) {
        const lastIndex = this.items.length - 1;
        item.durationObject.onCloseId = setTimeout(() => {
          this.onClose(item);
        }, duration);

        item.durationObject.style = {
          transition: `width ${duration / 1000}s`,
        };

        setTimeout(() => {
          item.durationObject.classes = ['hide'];

          this.$set(this.items, lastIndex, item);
        });
      }
    },

    onClose (closedItem: GToasterItem) {
      this.items = this.items.filter(item => item.id !== closedItem.id);

      this.executeCallback(closedItem, 'onCloseCallback');
    },

    hideAll () {
      this.items.forEach(item => {
        this.executeCallback(item, 'onCloseCallback');
      });
      this.items = [];
    },

    executeCallback (item: GToasterItem, type: CallbackType) {
      if (type === 'onCloseCallback' && item.closed) {
        return;
      }

      const defaultCallback = () => {};
      const callbackFn: (data: any) => void = item[type] || defaultCallback;

      const {
        onShowCallback,
        textColor,
        durationObject,
        variantObject,
        onCloseCallback,
        ...rest
      } = item;

      if (type === 'onCloseCallback') {
        item.closed = true;
      }

      callbackFn({ ...rest, closed: item.closed });
    },
  },
};

</script>

<style lang="scss" scoped>
  .toaster-wrapper {
    position: fixed;
    right: 10px;
    top: 10px;
    z-index: 999999999;

    .toaster {
      width: 325px;
      opacity: 0.9;
      border-radius: 2px;
      color: white;
      margin-top: 10px;
      position: relative;

      .header {
        font-weight: bold;
        padding-right: 25px;
        font-size: 15px;
        .title{
          display: block;
          padding: 5px 13px 0;
        }

        .close {
          cursor: pointer;
          position: absolute;
          top: 5px;
          right: 5px;
          font-size: 19px;
          color: white;
        }
      }

      .content {
        padding: 20px;
      }

      .text-color {
        color: var(--white);
      }

      &.error {
        background-color: var(--red-500);
        border-bottom: 1px solid var(--red-400);

        .duration-wrapper .duration {
          border-bottom: 3px solid var(--red-800);
        }
      }

      &.success {
        background-color: var(--green-500);
        border-bottom: 1px solid var(--green-400);
        .duration-wrapper .duration {
          border-bottom: 3px solid var(--green-800);
        }
      }

      &.warning {
        background-color: var(--orange-500);
        border-bottom: 1px solid var(--orange-400);
        .duration-wrapper .duration {
          border-bottom: 3px solid var(--orange-800);
        }
      }

      &.info {
        background-color: var(--blue-500);
        border-bottom: 1px solid var(--blue-400);
        color: #fff;
        .duration-wrapper .duration {
          border-bottom: 3px solid var(--blue-800);
        }
      }

      .duration-wrapper {
        position: absolute;
        left: 0;
        bottom: 0;
        width: 100%;
        .duration {
          width: 100%;

          &.hide {
            width: 0;
          }
        }
      }

    }

    .fade-enter-active, .fade-leave-active {
      transition: opacity .2s;
    }
    .fade-enter, .fade-leave-to {
      opacity: 0;
    }
  }
</style>
