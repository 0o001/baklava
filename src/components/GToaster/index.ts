import GToaster from './GToaster.vue';

export default GToaster;
