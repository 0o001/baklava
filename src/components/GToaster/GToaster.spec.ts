import { shallowMount, Wrapper } from '@vue/test-utils';
import GToaster from './';
import { GToasterItem } from '@/components/GToaster/types';

type GToasterType = {[key: string]: any}

describe('Toaster', () => {
  let wrapper: Wrapper<GToaster & GToasterType>;
  const renderWrapper = () => {
    wrapper = shallowMount(GToaster);
  };
  const clearTimeout = global.clearTimeout;

  beforeEach(() => {
    renderWrapper();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  afterAll(() => {
    global.clearTimeout = clearTimeout;
  });

  it('should match snapshot on initial state', () => {
    expect(wrapper.element)
      .toMatchSnapshot();
  });

  it('should match snapshot which is a success toaster, if config is not passed', () => {
    const successMessage = 'It is a warning';

    wrapper.vm.toast(successMessage);

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show variant name as a title, if title is not given', () => {
    wrapper.vm.toast('It is a test', {
      variant: 'error',
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should match snapshot which is as changed title and warning message, if title and warning variant is passed', () => {
    const warningMessage = 'It is a warning';

    wrapper.vm.toast(warningMessage, {
      title: 'Test Title',
      variant: 'warning',
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('show toasts as reversed', () => {
    wrapper.vm.toast('This is a error 1', {
      title: 'Test Title 1',
      variant: 'error',
    });
    wrapper.vm.toast('This is a error 2', {
      title: 'Test Title 2',
      variant: 'error',
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should not show same group toast if "unique" is passed as true', () => {
    wrapper.vm.toast('This is a error 1', {
      title: 'Test Title 1',
      group: 'testGroup',
      variant: 'error',
    });
    wrapper.vm.toast('This is a error 2', {
      title: 'Test Title 2',
      group: 'testGroup',
      unique: true,
      variant: 'error',
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should be disappeared after given duration time', async (done) => {
    wrapper.vm.toast('This is a test 1', {
      title: 'Test Title 1',
      duration: 50,
    });

    setTimeout(() => {
      expect(wrapper.element).toMatchSnapshot();
    }, 10);

    setTimeout(() => {
      expect(wrapper.find('.toaster').exists()).toBe(false);

      done();
    }, 55);
  });

  it('should be disappeared if click close button', async () => {
    wrapper.vm.toast('This is a test 1', {
      title: 'Test Title 1',
    });

    await wrapper.vm.$nextTick();
    wrapper.find('.close').trigger('click');
    await wrapper.vm.$nextTick();

    expect(wrapper.find('.toaster').exists()).toBe(false);
  });

  it('should clear all auto close timeout before component destroy', () => {
    global.clearTimeout = jest.fn();

    wrapper.vm.toast('This is a test 1', {
      title: 'Test Title 1',
      duration: 50,
    });

    wrapper.destroy();

    expect(global.clearTimeout).toBeCalled();
  });

  it('should close on click if the closeOnClick flag is set to true.', async () => {
    wrapper.vm.toast('This is a test 1', {
      title: 'Test Title 1',
    });

    await wrapper.vm.$nextTick();
    wrapper.find('.toaster').trigger('click');
    await wrapper.vm.$nextTick();

    expect(wrapper.find('.toaster').exists()).toBe(false);
  });

  it('should be able to render a component when a vue instance is provided.', async () => {
    wrapper.vm.toast({ name: 'this-is-a-test-component', template: '<p>This is a test 1</p>' }, {
      title: 'Test Title 1',
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.find('this-is-a-test-component-stub').exists()).toBe(true);
  });

  it('should hide all toasters when hide all called.', async () => {
    wrapper.vm.toast('This is a test 1', {
      title: 'Test Title 1',
    });
    await wrapper.vm.$nextTick();
    wrapper.vm.toast('This is a test 2', {
      title: 'Test Title 2',
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.findAll('.toaster').length).toEqual(2);

    wrapper.vm.hideAll();
    await wrapper.vm.$nextTick();

    expect(wrapper.findAll('.toaster').length).toEqual(0);
  });

  it('should call given onShowCallback after showing message', async () => {
    const onShowCallback = jest.fn();
    const messageItem: GToasterItem = {
      closeOnClick: true,
      closed: false,
      durationObject: { autoClose: false, seconds: 2000 },
      group: 'default',
      id: 0,
      message: 'This is a test 1',
      props: null,
      title: 'Test Title 1',
      unique: undefined,
      variantObject: { className: 'success' },
    };
    const { textColor, variantObject, durationObject, ...restItem } = messageItem;

    wrapper.vm.toast('This is a test 1', {
      title: 'Test Title 1',
      onShowCallback,
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.findAll('.toaster').length).toEqual(1);

    expect(onShowCallback).toBeCalledWith(restItem);
  });

  it('should call given onCloseCallback after closing message', async () => {
    const onCloseCallback = jest.fn();
    const messageItem: GToasterItem = {
      closeOnClick: true,
      closed: true,
      durationObject: { autoClose: false, seconds: 2000 },
      group: 'default',
      id: 0,
      message: 'This is a test 1',
      props: null,
      title: 'Test Title 1',
      unique: undefined,
      variantObject: { className: 'success' },
    };
    const { textColor, variantObject, durationObject, ...restItem } = messageItem;

    wrapper.vm.toast('This is a test 1', {
      title: 'Test Title 1',
      onCloseCallback,
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.toaster').length).toEqual(1);
    wrapper.vm.hideAll();
    await wrapper.vm.$nextTick();

    expect(wrapper.findAll('.toaster').length).toEqual(0);

    expect(onCloseCallback).toBeCalledWith(restItem);
  });

  it('should execute onCloseCallback if message is not closed', () => {
    const item: GToasterItem = {
      id: 1,
      closed: false,
      onCloseCallback: jest.fn(),
      durationObject: { seconds: 0, autoClose: false },
    };

    wrapper.vm.executeCallback(item, 'onCloseCallback');

    expect(item.onCloseCallback).toBeCalledWith({ id: 1, closed: true });
  });

  it('should not execute onCloseCallback if message is closed', () => {
    const item: GToasterItem = {
      id: 1,
      closed: true,
      onCloseCallback: jest.fn(),
      durationObject: { seconds: 0, autoClose: false },
    };

    wrapper.vm.executeCallback(item, 'onCloseCallback');

    expect(item.onCloseCallback).not.toBeCalled();
  });

  it('should execute onShowCallback', () => {
    const item: GToasterItem = {
      id: 1,
      closed: false,
      onShowCallback: jest.fn(),
      durationObject: { seconds: 0, autoClose: false },
    };

    wrapper.vm.executeCallback(item, 'onShowCallback');

    expect(item.onShowCallback).toBeCalledWith({ closed: false, id: 1 });
  });
});
