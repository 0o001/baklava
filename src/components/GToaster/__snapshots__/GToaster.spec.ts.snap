// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster should be disappeared after given duration time 1`] = `
<div
  class="toaster-wrapper"
>
  <transition-group-stub
    name="fade"
  >
    <div
      class="toaster success"
    >
      <div
        class="success_header header"
      >
        <div
          class="title"
        >
          <gtext-stub
            color="text-white"
            variant="body"
          >
            
            Test Title 1
          
          </gtext-stub>
        </div>
         
        <div
          class="close"
        >
          
          Ã—
        
        </div>
      </div>
       
      <div
        class="content"
      >
        <gtext-stub
          color="text-white"
          variant="body"
        >
          
          This is a test 1
        
        </gtext-stub>
      </div>
       
      <div
        class="duration-wrapper"
      >
        <div
          class="duration hide"
          style="transition: width 0.05s;"
        />
      </div>
    </div>
  </transition-group-stub>
</div>
`;

exports[`Toaster should match snapshot on initial state 1`] = `
<div
  class="toaster-wrapper"
>
  <transition-group-stub
    name="fade"
  />
</div>
`;

exports[`Toaster should match snapshot which is a success toaster, if config is not passed 1`] = `
<div
  class="toaster-wrapper"
>
  <transition-group-stub
    name="fade"
  />
</div>
`;

exports[`Toaster should match snapshot which is as changed title and warning message, if title and warning variant is passed 1`] = `
<div
  class="toaster-wrapper"
>
  <transition-group-stub
    name="fade"
  />
</div>
`;

exports[`Toaster should not show same group toast if "unique" is passed as true 1`] = `
<div
  class="toaster-wrapper"
>
  <transition-group-stub
    name="fade"
  />
</div>
`;

exports[`Toaster should show variant name as a title, if title is not given 1`] = `
<div
  class="toaster-wrapper"
>
  <transition-group-stub
    name="fade"
  />
</div>
`;

exports[`Toaster show toasts as reversed 1`] = `
<div
  class="toaster-wrapper"
>
  <transition-group-stub
    name="fade"
  />
</div>
`;
